From c5fe1b7f0846f9ae4898bd275c1b133ca310c9f4 Mon Sep 17 00:00:00 2001
From: Leo Shen <i@szclsya.me>
Date: Wed, 9 Apr 2025 11:07:24 -0400
Subject: [PATCH 2/3] qnx: add Standardized Layout for QNX System suport

- add $PREFIX and $FORBIDDEN_PREFIX, which is usually set as:
  - $PREFIX: /system
  - $FORBIDDEN_PREFIX: /usr
---
 abuild.in       | 169 +++++++++++++++++++++++++-----------------------
 functions.sh.in |  13 +++-
 2 files changed, 97 insertions(+), 85 deletions(-)

diff --git a/abuild.in b/abuild.in
index 587ef33..11dc37f 100644
--- a/abuild.in
+++ b/abuild.in
@@ -29,6 +29,12 @@ fi
 : ${ADDUSER:="abuild-adduser"}
 : ${ADDGROUP:="abuild-addgroup"}
 
+# Internal variable
+rprefix="${PREFIX#\/}"
+if [ $rprefix != "usr" ]; then
+    msg "Using non-FHS prefix (/$rprefix)"
+fi
+
 apk_opt_wait="--wait 30"
 doc_threshold=$((2 * 1024 * 1024)) # 2 MiB
 
@@ -780,9 +786,9 @@ targz() {
 postcheck() {
 	local dir="$1" name="$2" i= j= e=0
 	msg "Running postcheck for $name"
-	# checking for FHS compat
-	if ! options_has "!fhs"; then
-		for i in srv usr/local opt tmp var/tmp var/lock var/empty home sys proc mnt dev; do
+    # checking Standardized Layout for QNX Systems (SLQS) compat
+	if ! options_has "!slqs"; then
+		for i in $FORBIDDEN_PREFIX srv opt tmp var/tmp var/lock var/empty home sys proc mnt dev; do
 			for j in "$dir"/"$i"/* "$dir"/"$i"/.[!.]* "$dir"/"$i"/..?*; do
 				if [ -L "$j" ] || [ -e "$j" ]; then
 					error "Packages must not put anything under /$i"
@@ -791,10 +797,6 @@ postcheck() {
 				fi
 			done
 		done
-		if [ -d "$dir"/usr/var ]; then
-			error "Found /usr/var, localstatedir is most likely wrong"
-			e=1
-		fi
 	fi
 
 	# Alpine Linux as a musl libc distro does not use /lib64 or /usr/lib64 under
@@ -803,10 +805,10 @@ postcheck() {
 	# to warrant a lib64 check
 	if ! options_has "lib64"; then
 		if [ -e "$dir"/lib64 ]; then
-			error "Packages must not put anything under /lib64, use /usr/lib instead"
+			error "Packages must not put anything under /lib64, use $rprefix/lib instead"
 			e=1
-		elif [ -e "$dir"/usr/lib64 ]; then
-			error "Packages must not put anything under /usr/lib64, use /usr/lib instead"
+		elif [ -e "$dir"/"$rprefix"/lib64 ]; then
+			error "Packages must not put anything under $rprefix/lib64, use $rprefix/lib instead"
 			e=1
 		fi
 	fi
@@ -815,7 +817,7 @@ postcheck() {
 	if ! options_has "!usrmerge"; then
 		for i in bin sbin lib; do
 			if [ -L "$dir"/$i ] || [ -e "$dir"/$i ]; then
-				warning "Packages must not put anything under /$i, use /usr/${i##*s} instead"
+				warning "Packages must not put anything under /$i, use $rprefix/${i##*s} instead"
 			fi
 		done
 	fi
@@ -826,22 +828,22 @@ postcheck() {
 	fi
 
 	# look for /usr/share/doc
-	if [ -e "$dir"/usr/share/doc ] \
+	if [ -e "$dir"/"$rprefix"/share/doc ] \
 			&& ! is_doc_pkg "$name"; then
-		warning "Found /usr/share/doc but package name doesn't end with -doc"
+		warning "Found /$rprefix/share/doc but package name doesn't end with -doc"
 	fi
 	# look for /usr/share/devhelp
-	if [ -e "$dir"/usr/share/devhelp ] \
+	if [ -e "$dir"/"$rprefix"/share/devhelp ] \
 			&& ! is_devhelp_pkg "$name"; then
-		warning "Found /usr/share/devhelp but package name doesn't end with -devhelp"
+		warning "Found /$rprefix/share/devhelp but package name doesn't end with -devhelp"
 	fi
 	# look for /usr/share/man
-	if [ -e "$dir"/usr/share/man ]; then
+	if [ -e "$dir"/"$rprefix"/share/man ]; then
 		if ! is_doc_pkg "$name"; then
-			warning "Found /usr/share/man but package name doesn't end with -doc"
+			warning "Found /$rprefix/share/man but package name doesn't end with -doc"
 		fi
 		# check for uncompressed man pages
-		i=$(find "$dir"/usr/share/man -name '*.[0-8]' -type f | sed -e 's/^/\t/')
+		i=$(find "$dir"/"$rprefix"/share/man -name '*.[0-8]' -type f | sed -e 's/^/\t/')
 		if [ -n "$i" ]; then
 			error "Found uncompressed man pages:"
 			echo "$i"
@@ -851,7 +853,7 @@ postcheck() {
 	# look for pycache
 	# wildcard should always get the system python dir, and this is faster than
 	# trying to calculate the python version.
-	i="$(find "$dir"/usr/lib/python* \( -type d -a -name "__pycache__" \) 2>/dev/null || :)"
+	i="$(find $dir/$rprefix/lib/python* \( -type d -a -name "__pycache__" \) 2>/dev/null || :)"
 	if [ -n "$i" ] && [ "${name%-pyc}" = "$name" ]; then
 		warning "Found __pycache__ but package name doesn't end with -pyc"
 	fi
@@ -922,25 +924,25 @@ prepare_subpackages() {
 	done
 	postcheck "$pkgdir" "$pkgname" || return 1
 	# post check for /usr/share/locale
-	if [ -d "$pkgdir"/usr/share/locale ]; then
-		warning "Found /usr/share/locale"
+	if [ -d "$pkgdir"/"$rprefix"/share/locale ]; then
+		warning "Found /$rprefix/share/locale"
 		warning2 "Maybe add \$pkgname-lang to subpackages?"
 	fi
 	# post check for shell completions
-	if [ -d "$pkgdir"/usr/share/bash-completion ]; then
-		warning "Found /usr/share/bash-completion"
+	if [ -d "$pkgdir"/"$rprefix"/share/bash-completion ]; then
+		warning "Found /$rprefix/share/bash-completion"
 		warning2 "Add \$pkgname-bash-completion to subpackages"
 	fi
-	if [ -d "$pkgdir"/usr/share/zsh/site-functions ]; then
-		warning "Found /usr/share/zsh/site-functions"
+	if [ -d "$pkgdir"/"$rprefix"/share/zsh/site-functions ]; then
+		warning "Found /$rprefix/share/zsh/site-functions"
 		warning2 "Add \$pkgname-zsh-completion to subpackages"
 	fi
-	if [ -d "$pkgdir"/usr/share/fish/completions ]; then
-		warning "Found /usr/share/fish/completions"
-		warning2 "fish completions for programs should be located in /usr/share/fish/vendor_completions.d"
+	if [ -d "$pkgdir"/"$rprefix"/share/fish/completions ]; then
+		warning "Found /$rprefix/share/fish/completions"
+		warning2 "fish completions for programs should be located in /$rprefix/share/fish/vendor_completions.d"
 	fi
-	if [ -d "$pkgdir"/usr/share/fish/vendor_completions.d ]; then
-		warning "Found /usr/share/fish/completions"
+	if [ -d "$pkgdir"/"$rprefix"/share/fish/vendor_completions.d ]; then
+		warning "Found /$rprefix/share/fish/completions"
 		warning2 "Add \$pkgname-fish-completion to subpackages"
 	fi
 }
@@ -950,7 +952,7 @@ default_lang() {
 	install_if="$pkgname=$pkgver-r$pkgrel lang"
 	depends="$depends_lang"
 
-	amove ${langdir:-/usr/share/locale}
+	amove "${langdir:-/$rprefix/share/locale}"
 }
 
 lang() {
@@ -995,17 +997,20 @@ check_maintainer() {
 
 check_license() {
 	local ret=0
-	local license_list=/usr/share/spdx/license.lst
+	local license_lists="/$rprefix/share/spdx/license.lst /usr/share/spdx/license.lst"
 	if options_has "!spdx" || ! [ -f "$license_list" ]; then
+        warning "SPDX license file not available"
 		return 0
 	fi
 	local IFS="$IFS()"
 	local i; for i in $license; do
 		list_has "$i" AND OR WITH && continue
-		if ! grep -q -x -F "$i" "$license_list"; then
-			ret=1
-			warning "\"$i\" is not a known license"
-		fi
+        for license_list in $license_lists; do
+            if ! grep -q -x -F "$i" "$license_list"; then
+                ret=1
+                warning "\"$i\" is not a known license"
+            fi
+        done
 	done
 	return $ret
 }
@@ -1242,9 +1247,9 @@ prepare_trace_rpaths() {
 			printf("%s\n", $i);
 		}
 	}' | sort -u >"$controldir"/.rpaths
-	if grep -q -x '/usr/lib' "$controldir"/.rpaths; then
+	if grep -q -x "/$rprefix/lib" "$controldir"/.rpaths; then
 		# FIXME silence warning when $ORIGIN was used
-		warning "Redundant /usr/lib in rpath found"
+		warning "Redundant /$rprefix/lib in rpath found"
 	fi
 	if grep '^/home/' "$controldir"/.rpaths; then
 		error "Has /home/... in rpath"
@@ -1270,12 +1275,12 @@ prepare_pkgconfig_provides() {
 	local dir="${subpkgdir:-$pkgdir}" v
 	options_has "!tracedeps" && return 0
 	cd "$dir" || return 1
-	for i in usr/lib/pkgconfig/*.pc usr/share/pkgconfig/*.pc; do
+	for i in "$rprefix"/lib/pkgconfig/*.pc "$rprefix"/share/pkgconfig/*.pc; do
 		if ! [ -e "$i" ]; then
 			continue
 		fi
 		local f=${i##*/}
-		v=$(PKG_CONFIG_PATH="$dir"/usr/lib/pkgconfig:"$dir"/usr/share/pkgconfig \
+		v=$(PKG_CONFIG_PATH="$dir"/$rprefix/lib/pkgconfig:"$dir"/$rprefix/share/pkgconfig \
 			PKG_CONFIG_MAXIMUM_TRAVERSE_DEPTH=1 pkg-config \
 			--modversion ${f%.pc} | sed -E -e 's/-(alpha|beta|rc|pre)/_\1/')
 		v=${v#v}
@@ -1288,7 +1293,7 @@ prepare_command_provides() {
 	local dir="${subpkgdir:-$pkgdir}"
 	options_has "!tracedeps" && return 0
 	cd "$dir" || return 1
-	for i in bin/* sbin/* usr/bin/* usr/sbin/*; do
+	for i in bin/* sbin/* "$rprefix"/bin/* "$rprefix"/sbin/*; do
 		# We can't check symlinks properly with absolute paths,
 		# subpackages and dependencies so assume that they are commands
 		if ! [ -L "$i" ]; then
@@ -1316,7 +1321,7 @@ prepare_py_provides() {
 	# Find all directories under site-packages, ignore __pycache__ and
 	# .egg-info packages, store the basename with the proper pyX.Y prefix
 	# one per-line
-	py_providers="$(find 'usr/lib/python'*/site-packages \
+	py_providers="$(find $rprefix/lib/python*/site-packages \
 		-mindepth 1 -maxdepth 1 \
 		-type f -iname '*.py' -o -type d \
 		-exec sh -c '
@@ -1335,7 +1340,7 @@ prepare_py_provides() {
 		# Find out which version of python we are building for
 		# this will find out by looking at the MAJOR and MINOR
 		# versions in /usr/lib/pythonMAJOR.MINOR
-		pyver="${d##*usr/lib/python}"
+		pyver="${d##*$rprefix/lib/python}"
 		pyver="${pyver%%/*}"
 
 		# Strip the .py prefix if it exists
@@ -1427,7 +1432,7 @@ real_so_path() {
 	local so="$1"
 	shift
 	while [ $# -gt 0 ]; do
-		[ -e "$1"/$so ] && realpath "$1/$so" && return 0
+		[ -e "$1"/"$so" ] && realpath "$1/$so" && return 0
 		shift
 	done
 	error "$so: path not found"
@@ -1440,7 +1445,7 @@ find_so_files() {
 	rpaths=$(cat "$1")
 	shift
 	while [ $# -gt 0 ]; do
-		real_so_path "$1" /usr/lib /lib $rpaths || return 1
+		real_so_path "$1" /"$rprefix"/lib /lib $rpaths || return 1
 		shift
 	done
 	return 0
@@ -1478,7 +1483,7 @@ trace_apk_deps() {
 
 	msg "Tracing dependencies..."
 	# add pkgconfig if usr/lib/pkgconfig is found
-	if [ -d "$pkgbasedir"/$name/usr/lib/pkgconfig ] || [ -d "$pkgbasedir"/$name/usr/share/pkgconfig ] \
+	if [ -d "$pkgbasedir"/$name/$rprefix/lib/pkgconfig ] || [ -d "$pkgbasedir"/$name/$rprefix/share/pkgconfig ] \
 			&& ! grep -q '^depend = pkgconfig' "$dir"/.PKGINFO; then
 		autodeps="$autodeps pkgconfig"
 	fi
@@ -1512,7 +1517,7 @@ trace_apk_deps() {
 
 find_scanelf_paths() {
 	local datadir="$1"
-	local paths="$datadir/lib:$datadir/usr/lib" i= rpaths=
+	local paths="$datadir/lib:$datadir/$rprefix/lib" i= rpaths=
 	if [ -n "$ldpath" ]; then
 		paths="$paths:$(echo "${datadir}${ldpath}" | sed "s|:|:$datadir|g")"
 	fi
@@ -1656,7 +1661,7 @@ scan_python3_dependency() {
 	local controldir="$2" datadir="$3"
 	local dir_count=0
 	local site_pkg_dir
-	for site_pkg_dir in "$datadir"/usr/lib/python3*/site-packages; do
+	for site_pkg_dir in "$datadir"/"$rprefix"/lib/python3*/site-packages; do
 		if ! [ -d "$site_pkg_dir" ]; then
 			# empty iteration
 			continue
@@ -1666,7 +1671,7 @@ scan_python3_dependency() {
 			error "package contains python3 modules for conflicting python3 versions"
 			return 1
 		fi
-		local pyver="${site_pkg_dir##*usr/lib/python}"
+		local pyver="${site_pkg_dir##*"$rprefix"/lib/python}"
 		pyver="${pyver%%/*}"
 		if [ -n "$pyver" ] && [ "${subpkgname:-$pkgname}" != python3 ]; then
 			echo "python3~$pyver" \
@@ -1677,7 +1682,7 @@ scan_python3_dependency() {
 
 # find the packages containing the binaries corresponding to the debug files
 scan_debuginfo_targets() {
-	local root=$3/usr/lib/debug
+	local root=$3/"$rprefix"/lib/debug
 	[ -d "$root" ] || return 0
 
 	find "$root" -name "*.debug" | while read path; do
@@ -1699,7 +1704,7 @@ scan_pkgconfig_depends() {
 	name="$(pkginfo_val pkgname "$controldir"/.PKGINFO)"
 	datadir="$pkgbasedir"/$name
 	for i in $(sort -u "$provides_pc"); do
-		PKG_CONFIG_PATH="$datadir"/usr/lib/pkgconfig:"$datadir"/usr/share/pkgconfig \
+		PKG_CONFIG_PATH="$datadir"/$rprefix/lib/pkgconfig:"$datadir"/$rprefix/share/pkgconfig \
 				pkg-config \
 				--print-requires \
 				--print-requires-private ${i%=*} \
@@ -1725,10 +1730,10 @@ create_apks() {
 	local file= dir= name= ver= apk= datadir= size= i= gzip=
 	gzip=$(command -v pigz || echo gzip)
 	if ! options_has "keepdirs"; then
-		rmdir "$pkgdir"/usr/lib \
-			"$pkgdir"/usr/bin \
-			"$pkgdir"/usr/share \
-			"$pkgdir"/usr \
+		rmdir "$pkgdir"/"$rprefix"/lib \
+			"$pkgdir"/"$rprefix"/bin \
+			"$pkgdir"/"$rprefix"/share \
+			"$pkgdir"/"$rprefix" \
 			"$pkgdir"/etc/ \
 			"$pkgdir"/lib/ \
 			2>/dev/null || :
@@ -1899,11 +1904,11 @@ default_doc() {
 
 	local i
 	for i in doc man info html sgml licenses gtk-doc ri help; do
-		amove usr/share/"$i" || :
+		amove "$rprefix"/share/"$i" || :
 	done
 
 	# compress man pages
-	local mandir="$subpkgdir"/usr/share/man
+	local mandir="$subpkgdir"/"$rprefix"/share/man
 	[ -d "$mandir" ] && find "$mandir" -type l \
 		-a \( -name \*.[0-9n] -o -name \*.[0-9n][a-z]* \) \
 		-a \! \( -name '*.gz' -o -name '*.bz2' -o -name '*.xz' \) \
@@ -1935,7 +1940,7 @@ default_doc() {
 		[ $islink -eq 0 ] && $gzip -n -9 "$name"
 	done
 
-	rm -f "$subpkgdir/usr/share/info/dir"
+	rm -f "$subpkgdir/$rprefix/share/info/dir"
 }
 
 doc() {
@@ -1951,7 +1956,7 @@ default_dbg() {
 	mkdir -p "$pkgbasedir/.dbg-tmp"
 	# note: the subpkgdir doesn't exist when the glob is evaluated
 	scanelf -RyB -E ET_DYN "$pkgbasedir"/* | while read type src; do
-		local ino dst=$subpkgdir/usr/lib/debug/${src#"$pkgbasedir"/*/}.debug
+		local ino dst=$subpkgdir/$rprefix/lib/debug/${src#"$pkgbasedir"/*/}.debug
 		mkdir -p "${dst%/*}"
 		ino=$(stat -c %i "$src")
 		if ! [ -e "$pkgbasedir/.dbg-tmp/$ino" ]; then
@@ -1979,14 +1984,14 @@ default_dev() {
 	pkgdesc="$pkgdesc (development files)"
 
 	cd "$pkgdir" || return 0
-	local libdirs=usr/
+	local libdirs="$rprefix"/
 	[ -d lib/ ] && libdirs="lib/ $libdirs"
-	for i in usr/include usr/lib/pkgconfig usr/share/pkgconfig \
-			usr/share/aclocal usr/share/gettext \
-			usr/bin/*-config usr/share/vala/vapi \
-			usr/share/gir-[0-9]* usr/share/qt*/mkspecs \
-			usr/lib/qt*/mkspecs usr/lib/cmake usr/share/cmake \
-			usr/lib/glade/modules usr/share/glade/catalogs \
+	for i in $rprefix/include $rprefix/lib/pkgconfig $rprefix/share/pkgconfig \
+			$rprefix/share/aclocal $rprefix/share/gettext \
+			$rprefix/bin/*-config $rprefix/share/vala/vapi \
+			$rprefix/share/gir-[0-9]* $rprefix/share/qt*/mkspecs \
+			$rprefix/lib/qt*/mkspecs $rprefix/lib/cmake $rprefix/share/cmake \
+			$rprefix/lib/glade/modules $rprefix/share/glade/catalogs \
 			$(find . -name include -type d) \
 			$(subpackage_types_has static || find $libdirs \
 				-name '*.a' 2>/dev/null) \
@@ -1997,7 +2002,7 @@ default_dev() {
 		fi
 	done
 	# move *.so links needed when linking the apps to -dev packages
-	for i in lib/*.so usr/lib/*.so; do
+	for i in lib/*.so "$rprefix"/lib/*.so; do
 		[ -L "$i" ] && amove "$i"
 	done
 	return 0
@@ -2025,7 +2030,7 @@ default_static() {
 	pkgdesc="$pkgdesc (static library)"
 
 	cd "$pkgdir" || return 0
-	local libdirs=usr/lib
+	local libdirs="$rprefix"/lib
 	[ -d lib/ ] && libdirs="lib/ $libdirs"
 
 	# move *.a static library
@@ -2043,7 +2048,7 @@ default_libs() {
 	depends="$depends_libs"
 	pkgdesc="$pkgdesc (libraries)"
 
-	amove 'lib/lib*.so.[0-9]*' 'usr/lib/lib*.so.[0-9]*' || :
+	amove lib/'lib*.so.[0-9]*' "$rprefix"/lib/'lib*.so.[0-9]*' || :
 }
 
 libs() {
@@ -2068,7 +2073,7 @@ default_devhelp() {
 	pkgdesc="$pkgname - devhelp files"
 	install_if="$pkgname=$pkgver-r$pkgrel devhelp"
 
-	amove usr/share/devhelp
+	amove "$rprefix"/share/devhelp
 }
 
 devhelp() {
@@ -2081,7 +2086,7 @@ default_bashcomp() {
 	install_if="${subpkgname%-bash-completion}=$pkgver-r$pkgrel bash-completion"
 
 	cd "$pkgdir" || return 0
-	amove usr/share/bash-completion/completions
+	amove "$rprefix"/share/bash-completion/completions
 }
 
 bashcomp() {
@@ -2093,7 +2098,7 @@ default_zshcomp() {
 	pkgdesc="Zsh completions for $pkgname"
 	install_if="${subpkgname%-zsh-completion}=$pkgver-r$pkgrel zsh"
 
-	amove usr/share/zsh/site-functions
+	amove "$rprefix"/share/zsh/site-functions
 }
 
 zshcomp() {
@@ -2106,10 +2111,10 @@ default_fishcomp() {
 	install_if="${subpkgname%-fish-completion}=$pkgver-r$pkgrel fish"
 
 	cd "$pkgdir" || return 0
-	amove usr/share/fish/vendor_completions.d || :
+	amove "$rprefix"/share/fish/vendor_completions.d || :
 
-	amove usr/share/fish/completions &&
-		warning "fish completions for programs should be in /usr/share/fish/vendor_completions.d"
+	amove "$rprefix"/share/fish/completions &&
+		warning "fish completions for programs should be in /$rprefix/share/fish/vendor_completions.d"
 
 	return 0
 }
@@ -2124,7 +2129,7 @@ default_pyc() {
 
 	cd "$pkgdir" || return 0
 	local IFS=$'\n'
-	amove $(find usr/lib/python* -type d -name __pycache__)
+	amove $(find "$rprefix"/lib/python* -type d -name __pycache__)
 }
 
 pyc() {
@@ -2455,13 +2460,13 @@ rootbld() {
 		"$BUILD_ROOT/$ABUILD_USERDIR" "$BUILD_ROOT/$aportsgit" \
 		"$BUILD_ROOT/$SRCDEST" "$BUILD_ROOT/$REPODEST" \
 		"$BUILD_ROOT/tmp/pkg" "$BUILD_ROOT/tmp/src" \
-		"$BUILD_ROOT/tmp/tmp" "$BUILD_ROOT/usr/bin" \
+		"$BUILD_ROOT/tmp/tmp" "$BUILD_ROOT/$rprefix/bin" \
 		"$REPODEST" "$BUILD_ROOT/$HOME/.ccache"
 
 	cp /etc/abuild.conf "$BUILD_ROOT/etc"
 
 	local dir
-	for dir in /usr/share/apk/keys/$CBUILD_ARCH /etc/apk/keys; do
+	for dir in /$rprefix/share/apk/keys/$CBUILD_ARCH /etc/apk/keys; do
 		cp $dir/* "$BUILD_ROOT/etc/apk/keys"
 	done
 
@@ -2560,10 +2565,10 @@ rootbld() {
 		${APORTS_BOOTSTRAP:+--setenv APORTS_BOOTSTRAP "$APORTS_BOOTSTRAP"} \
 		--setenv SOURCE_DATE_EPOCH "$SOURCE_DATE_EPOCH" \
 		--setenv ABUILD_LAST_COMMIT "$ABUILD_LAST_COMMIT" \
-		--setenv PATH ${USE_CCACHE:+/usr/lib/ccache/bin:}/bin:/usr/bin:/sbin:/usr/sbin \
+		--setenv PATH ${USE_CCACHE:+/$rprefix/lib/ccache/bin:}/bin:/$rprefix/bin:/sbin:/$rprefix/sbin \
 		--setenv FAKEROOTDONTTRYCHOWN 1 \
 		--unsetenv CBUILD \
-		/usr/bin/abuild $force rootbld_actions
+		/$rprefix/bin/abuild $force rootbld_actions
 	update_abuildrepo_index
 	cleanup $CLEANUP
 }
@@ -2591,7 +2596,7 @@ stripbin() {
 		[ -e "$filename" ] || continue
 
 		case "$filename" in
-			./usr/lib/debug/*) continue ;;
+			./$rprefix/lib/debug/*) continue ;;
 		esac
 
 		[ "$osabi" != "STANDALONE" ] || continue
@@ -2863,7 +2868,7 @@ fi
 if [ -n "$USE_CCACHE" ]; then
 	mkdir -p "$HOME/.ccache"
 	# prepend ccache path
-	export PATH="/usr/lib/ccache/bin:$PATH"
+	export PATH="/$rprefix/lib/ccache/bin:$PATH"
 fi
 
 # find startdir
diff --git a/functions.sh.in b/functions.sh.in
index c0302a9..9835ef5 100644
--- a/functions.sh.in
+++ b/functions.sh.in
@@ -75,6 +75,7 @@ readconfig() {
 	local _JOBS _MAKEFLAGS _PACKAGER _USE_COLORS _USE_CCACHE
 	local _BUILDCC _BUILDCXX _BUILDLD _BUILDCPPFLAGS _BUILDCXXFLAGS _BUILDCFLAGS
 	local _BUILDLDFLAGS
+    local _PREFIX
 	local gitbase=
 	[ -n "${APORTSDIR+x}" ] && _APORTSDIR=$APORTSDIR
 	[ -n "${BUILDDIR+x}" ] && _BUILDDIR=$BUILDDIR
@@ -104,6 +105,7 @@ readconfig() {
 	[ -n "${PACKAGER+x}" ] && _PACKAGER=$PACKAGER
 	[ -n "${USE_COLORS+x}" ] && _USE_COLORS="$USE_COLORS"
 	[ -n "${USE_CCACHE+x}" ] && _USE_CCACHE="$USE_CCACHE"
+    [ -n "${PREFIX+x}" ] && _PREFIX=$PREFIX
 	: ${ABUILD_DEFCONF:=$sharedir/default.conf}
 	[ -f "$ABUILD_DEFCONF" ] && . "$ABUILD_DEFCONF"
 	: ${ABUILD_CONF:=$sysconfdir/abuild.conf}
@@ -159,6 +161,11 @@ readconfig() {
 	PACKAGER=${_PACKAGER-$PACKAGER}
 	USE_COLORS=${_USE_COLORS-$USE_COLORS}
 	USE_CCACHE=${_USE_CCACHE-$USE_CCACHE}
+    PREFIX=${_PREFIX-$PREFIX}
+    if [ "$PREFIX" = "${PREFIX#\/}" ]; then
+        echo "\$PREFIX is relative. It must be absolute!"
+        exit 1
+    fi
 
 	# export BUILD variables because it is used as variables specific to the platform running the compiler
 	# instead of the platform the compiled binaries are supposed to run on
@@ -194,11 +201,11 @@ readconfig() {
 		# setup build root
 		[ -z "$CBUILDROOT" ] && export CBUILDROOT="$HOME/sysroot-$CTARGET_ARCH/"
 		# prepare pkg-config for cross building
-		[ -z "$PKG_CONFIG_PATH" ] && export PKG_CONFIG_PATH="${CBUILDROOT}/usr/lib/pkgconfig/:${CBUILDROOT}/usr/share/pkgconfig/"
+		[ -z "$PKG_CONFIG_PATH" ] && export PKG_CONFIG_PATH="${CBUILDROOT}/${PREFIX}/lib/pkgconfig/:${CBUILDROOT}/${PREFIX}/share/pkgconfig/"
 		[ -z "$PKG_CONFIG_SYSROOT_DIR" ] && export PKG_CONFIG_SYSROOT_DIR="${CBUILDROOT}"
 		# libtool bug workaround for extra rpaths
 		[ -z "$lt_cv_sys_lib_dlsearch_path_spec" ] && \
-			export lt_cv_sys_lib_dlsearch_path_spec="${CBUILDROOT}/lib ${CBUILDROOT}/usr/lib /usr/lib /lib /usr/local/lib"
+			export lt_cv_sys_lib_dlsearch_path_spec="${CBUILDROOT}/lib ${CBUILDROOT}/${PREFIX}/lib /${PREFIX}/lib /lib /${PREFIX}/local/lib"
 		# setup cross-compiler
 		if [ -z "$CROSS_COMPILE" ]; then
 			export CROSS_COMPILE="${CHOST}-"
@@ -215,7 +222,7 @@ readconfig() {
 			export CPPFLAGS="--sysroot=${CBUILDROOT} $CPPFLAGS"
 			export CXXFLAGS="--sysroot=${CBUILDROOT} $CXXFLAGS"
 			export CFLAGS="--sysroot=${CBUILDROOT} $CFLAGS"
-			export LDFLAGS="--sysroot=${CBUILDROOT} $LDFLAGS"
+			export LDFLAGS="--sysroot=${CBUILDROOT} -L$CBUILDROOT/$PREFIX/lib $LDFLAGS"
 		fi
 	fi
 	return 0
-- 
2.34.1

